File: Source/Bifrost/Utils/StringMapper.cs
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) 2008-2017 Dolittle. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
using System.Collections.Generic;
using System.Linq;

namespace Bifrost.Utils
{
    /// <summary>
    /// Represents an implementation of <see cref="IStringMapper"/>
    /// </summary>
    public class StringMapper : IStringMapper
    {
        List<IStringMapping> _mappings = new List<IStringMapping>();

#pragma warning disable 1591 // Xml Comments
        public IEnumerable<IStringMapping> Mappings => _mappings;

        public bool HasMappingFor(string input)
        {
            return Mappings.Any(mapping => mapping.Matches(input));
        }

        public IStringMapping GetFirstMatchingMappingFor(string input)
        {
            return Mappings.FirstOrDefault(mapping => mapping.Matches(input));
        }

        public IEnumerable<IStringMapping> GetAllMatchingMappingsFor(string input)
        {
            return Mappings.Where(mapping => mapping.Matches(input));
        }

        public string Resolve(string input)
        {
            return GetFirstMatchingMappingFor(input)?.Resolve(input);
        }

        public void AddMapping(string format, string mappedFormat)
        {
            _mappings.Add(new StringMapping(format, mappedFormat));
        }

        public void AddMapping(IStringMapping mapping)
        {
            _mappings.Add(mapping);
        }
#pragma warning restore 1591 // Xml Comments
    }
}

File: Source/Bifrost.JavaScript/utils/StringMapper.js
Bifrost.namespace("Bifrost", {
    StringMapper: Bifrost.Type.extend(function (stringMappingFactory) {
        var self = this;

        this.stringMappingFactory = stringMappingFactory;

        this.mappings = [];

        this.reverseMappings = [];

        function hasMappingFor(mappings, input) {
            var found = false;
            mappings.some(function (m) {
                if (m.matches(input)) {
                    found = true;
                }
                return found;
            });
            return found;
        }

        function getMappingFor(mappings, input) {
            var found;
            mappings.some(function (m) {
                if (m.matches(input)) {
                    found = m;
                    return true;
                }
            });

            if (typeof found !== "undefined") {
                return found;
            }

            throw {
                name: "ArgumentError",
                message: "String mapping for (" + input + ") could not be found"
            };
        }

        function resolve(mappings, input) {
            try {
                if (input === null || typeof input === "undefined") {
                    return "";
                }
                
                var mapping = self.getMappingFor(input);
                if (Bifrost.isNullOrUndefined(mapping)) {
                    return "";
                }

                return mapping.resolve(input);
            } catch (e) {
                return "";
            }
        }

        this.hasMappingFor = function (input) {
            return hasMappingFor(self.mappings, input);
        };
        this.getMappingFor = function (input) {
            return getMappingFor(self.mappings, input);
        };
        this.resolve = function (input) {
            return resolve(self.mappings, input);
        };

        this.reverse = {
            hasMappingFor: function (input) {
                return self.hasMappingFor(self.reverseMappings, input);
            },

            getMappingFor: function (input) {
                return self.getMappingFor(self.reverseMappings, input);
            },

            resolve: function (input) {
                return self.resolve(self.reverseMappings, input);
            }
        };

        this.addMapping = function (format, mappedFormat) {
            var mapping = self.stringMappingFactory.create(format, mappedFormat);
            self.mappings.push(mapping);

            var reverseMapping = self.stringMappingFactory.create(mappedFormat, format);
            self.reverseMappings.push(reverseMapping);
        };
    })
});